pyquil/quilatom.py

```diff
@@ -899,26 +900,58 @@ def out(self) -> str:
         return str(self)
 
 
-@dataclass
-class WaveformReference(QuilAtom):
+class WaveformReference(quil_rs.WaveformInvocation):
     """
     Representation of a Waveform reference.
     """
 
-    name: str
-    """ The name of the waveform. """
+    def __new__(cls, name: str):
+        return super().__new__(cls, name, {})
 
     def out(self) -> str:
-        return self.name
+        return str(self)
+
+
+def _template_waveform_property(name: str, doc: Optional[str] = None):
+    """
+    Helper method for initializing getters, setters, and deleters for
+    parameters on a ``TemplateWaveform``. Should only be used inside of
+    ``TemplateWaveform`` or one its base classes.
+    """
+
+    def fget(self: "TemplateWaveform"):
+        return self.get_parameter(name)
+
+    def fset(self: "TemplateWaveform", value: ParameterDesignator):
+        self.set_parameter(name, value)
+
+    def fdel(self: "TemplateWaveform"):
+        self.set_parameter(name, None)
+
+    return property(fget, fset, fdel, doc)
 
-    def __str__(self) -> str:
-        return self.out()
 
+class TemplateWaveform(quil_rs.WaveformInvocation, QuilAtom):
+    def __new__(cls, name: str, *, duration: float, **kwargs):
+        rs_parameters = {key: _convert_to_rs_expression(value) for key, value in kwargs.items() if value is not None}
+        rs_parameters["duration"] = _convert_to_rs_expression(duration)
+        return super().__new__(cls, name, rs_parameters)
 
-@dataclass
-class TemplateWaveform(QuilAtom):
-    duration: float
-    """ The duration [seconds] of the waveform. """
+    def out(self) -> str:
+        return str(self)
+
+    def get_parameter(self, name: str) -> Optional[ParameterDesignator]:
+        return _convert_to_py_parameter(super().parameters.get(name, None))
+
+    def set_parameter(self, name: str, value: Optional[ParameterDesignator]):
+        parameters = super().parameters
+        if value is None:
+            parameters.pop(name, None)
+        else:
+            parameters[name] = _convert_to_rs_expression(value)
+        quil_rs.WaveformInvocation.parameters.__set__(self, parameters)
+
+    duration = _template_waveform_property("duration")
```

Comment:
Don't love the duplication of the name here, and there probably is a way to decorate it such that it isn't necessary, but couldn't find a clean way to do it.